### DISCUSSION TITLE
Historic queries & link properties

### DISCUSSION THREAD

**Discussion by adriangb** [geldata/gel#1991] [Q&A] [answered]
First off thank you for the great project! I've been looking for something like this for a while.

Edit: removed recursive query question since it's answered in #1764.

I need to be able to query the state of my DB at any point in time. In a SQL DB, this would be solved with audit tables. In a graph DB, it could be done using properties on links/nodes ([example](https://www.arangodb.com/2018/07/time-traveling-with-graph-databases/)). In EdgeDB, I thought of doing something like this:
```
abstract link timed_link {
    property created_at -> datetime;
    property deleted_at -> datetime;
}
type Container {
    required property identifier -> str {
        constraint exclusive;
    };
    multi link contains extending timed_link -> Item;
}
```
I have not been able to figure out the EdgeQL syntax to connect a `Container` with its `contains`. I tried stuff along the following:
```
INSERT Container {
    identifier := 'container12345',
};
UPDATE Container
FILTER .identifier = 'container12345'
SET {
    contains := (
        SELECT Item
        FILTER Item.identifier = 'item12345'
    ) {
        create_time := datetime_of_transaction(),
    }
}

```
So my questions are:
* **What is the EdgeQL syntax to set a property of a link to a given value?**
* **Is there any better way to maintain historical state**

**Comment by elprans** [comment, marked as answer]
Use the `@`-sign to refer to link properties, like so:

```
UPDATE Container
FILTER .identifier = 'container12345'
SET {
    contains := (
        SELECT Item
        FILTER Item.identifier = 'item12345'
    ) {
        @created_at := datetime_of_transaction(),
    }
}
```

**Comment by elprans** [comment]
> Is there any better way to maintain historical state

Link properties would be one way.  Another is to incorporate the state into the subordinate objects, reverse the link relationship and then use computable links for convenience.

Suppose you have a CMS where each post can have multiple revisions.  We can represent that with a schema like this:

```
type Revision {
    required property created_at -> datetime {
        constraint exclusive;
    }
    required property modified_at -> datetime;
    required property published -> bool {
        default := false;
    }
    required property body -> str;
    required link post -> Post;
};

type Post {
    required property slug -> str;
    link published_revision := (
        # backward link navigation to resolve related revisions
        SELECT .<post[IS Revision]
        # only revisions that aren't in draft
        FILTER .published
        # get the latest
        ORDER BY .created_at DESC
        LIMIT 1
    );
};
```

Then, for the most common case of rendering the latest revision you'd just do

```
    SELECT Post {
        published_revision: {
            body
        }
    }
    FILTER .slug = '/foo/bar';
```

And for the case of rendering a specific revision you could restate the revision query:

```
    WITH 
        MyPost := Post {
            specific_revision := (
                SELECT .<post[IS Revision]
                FILTER .created_at = <datetime>$desired_version
            )
        }
    SELECT MyPost {
        specific_revision: {
            body
        }
    }
    FILTER .slug = '/foo/bar';
```

**Reply by adriangb** [reply]
This is great. I was indeed also testing with a `VersionNode` concept that sits in the middle of links, similar to what you propose, but it would not have been as clean and it would have taken me ages to figure out the queries, so thank you very much for expressing the idea so cleanly.

Aside from being able to make the timestamps required, it is also nice because link properties are not a very universal concept. That is, vertex properties are a pretty basic thing in any type of graph thingy, edge properties -aside from weights- are not that universal (I think).

**Comment by adriangb** [comment]
I'm still having a bit of trouble with syntax... Working with this schema:
```
abstract link timed_link {
    property deleted_at -> datetime;
}
type Container {
    required property name -> str;
    multi link contains_history extending timed_link -> Container;
    multi link contains := (
        SELECT .contains_history
        FILTER @deleted_at >= datetime_of_transaction()
    );
}
```
Where I am trying to declare a computable multi link named `contains` that gathers all of the _current_ links from `contains_history`.
I populate my data like so:
```
INSERT Container { name := "Container1" };
INSERT Container { name := "Container2" };
INSERT Container {
    name := "Container3",
    contains_history := (
        SELECT DETACHED Container
    ) {
        @deleted_at := to_datetime('2030-05-07T15:01:22.306916+00')
    }
};
```
Which creates 3 containers with the 3rd (`Container3`) "containing" the first two.

Now if I want to query all of the containers contained by `Container3`, the following works:
```
SELECT Container.contains_history
FILTER .<contains_history[IS Container].name = "Container3";
```
But if I want to use the `contains` computable link that I defined:
```
SELECT Container.contains
FILTER .<contains[IS Container].name = "Container3";
```
I get:
```
reference to 'Container' changes the interpretation of 'Container' elsewhere in the query
```
I presume this is related to #977, but I am having trouble figuring out what that implies in this context.

**Reply by elprans** [reply]
This seems to be a bug.  Your query appears to be valid.  Will look into a fix soon.

As for the "interpretation" error in general, I wrote a bit of an explanation here: https://github.com/edgedb/edgedb/discussions/1854#discussioncomment-89672

**Reply by adriangb** [reply]
Got it, thank you for taking a look so quickly!

**Reply by adriangb** [reply]
I'm also getting a similar error with:
```
START MIGRATION TO {
    module default {
        abstract link timed_link {
            property deleted_at -> datetime;
        }
        type Container {
            required property name -> str;
            multi link contains_history extending timed_link -> Container;
            multi link contains := (
                SELECT .contains_history
                FILTER @deleted_at >= datetime_of_transaction()
            );
        }
        type ConstrainedContainer extending Container {
            constraint expression on (
                count(.contains) < 10
            )
        }
    }
};
```
```
reference to 'ConstrainedContainer' changes the interpretation of 'ConstrainedContainer' elsewhere in the query
  ┌─ query:2:13
  │  
2 │       module default {
  │ ╭─────────────^
3 │ │         abstract link timed_link {
  │ ╰───^ error
```

**Reply by elprans** [reply]
Yep, it's the same bug.
